src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+         }
+      }
+      
+      const auto processResult = audioProcessor->process(data);
+      
+      return processResult == kResultOk ?
+         data.numSamples : 0;
+   }
+   wxLogError("Failed to query IAudioProcess interface from the effect component");
+   return 0;
+   }
+}
+
+size_t VST3Effect::ProcessBlock(float** inBlock, float** outBlock, size_t blockLen)
+{
+   internal::ComponentHandler::PendingChangesPtr pendingChanges { nullptr };
+   if(mComponentHandler)
+      pendingChanges = mComponentHandler->getPendingChanges();
+   return VST3ProcessBlock(mEffectComponent.get(), mSetup, inBlock, outBlock, blockLen, pendingChanges.get());
+}
+
+bool VST3Effect::RealtimeInitialize()
+{
+   //reload current parameters form the editor into parameter queues
+   SyncParameters();
+   return true;
+}
+
+bool VST3Effect::RealtimeAddProcessor(unsigned numChannels, float sampleRate)
+{
+   using namespace Steinberg;
+
+   auto effect = std::make_unique<VST3Effect>(*this);
+   effect->mSetup.processMode = Vst::kRealtime;
+   effect->mSetup.sampleRate = sampleRate;
+   
+   if(!effect->ProcessInitialize({0}, nullptr))
+   {
+      wxLogError("VST3 realtime initialization failed");
+      return false;
+   }
+
+   mRealtimeGroupProcessors.push_back(std::move(effect));
+   return true;
+}
+
+bool VST3Effect::RealtimeFinalize()
+{
+   for(auto& processor : mRealtimeGroupProcessors)
+      processor->ProcessFinalize();
+
+   mRealtimeGroupProcessors.clear();
+   mPendingChanges.reset();
+   
+   return true;
+}
+
+bool VST3Effect::RealtimeSuspend()
+{
+   using namespace Steinberg;
+   for(auto& effect : mRealtimeGroupProcessors)
+   {
+      Vst::IAudioProcessor* audioProcessor;
+      if(effect->mEffectComponent->queryInterface(Vst::IAudioProcessor::iid, reinterpret_cast<void**>(&audioProcessor)) == kResultOk)
+         audioProcessor->setProcessing(false);
+   }
+   return true;
+}
+
+bool VST3Effect::RealtimeResume()
+{
+   using namespace Steinberg;
+   for(auto& effect : mRealtimeGroupProcessors)
+   {
+      Vst::IAudioProcessor* audioProcessor;
+      if(effect->mEffectComponent->queryInterface(Vst::IAudioProcessor::iid, reinterpret_cast<void**>(&audioProcessor)) == kResultOk)
+         audioProcessor->setProcessing(true);
+   }
+   return true;
+}
+
+bool VST3Effect::RealtimeProcessStart()
+{
+   mPendingChanges = mComponentHandler->getPendingChanges();
+   return true;
+}
+
+size_t VST3Effect::RealtimeProcess(int group, float** inBuf, float** outBuf, size_t numSamples)
+{
+   auto& effect = mRealtimeGroupProcessors[group];
+   return VST3ProcessBlock(effect->mEffectComponent.get(), effect->mSetup, inBuf, outBuf, numSamples, mPendingChanges.get());
+}
+
+bool VST3Effect::RealtimeProcessEnd()
+{
+   mPendingChanges.reset();
+   return true;
+}
+
+int VST3Effect::ShowClientInterface(wxWindow& parent, wxDialog& dialog, bool forceModal)
+{
+   dialog.Show();
+   return 0;
+}
+
+bool VST3Effect::SetHost(EffectHostInterface* host)
+{
+   //Dummy
```

Comment:
Is this TODO?

Other overrides re-load last used settings from saved preferences.  This function is called the first time the effect is used again in a session.